### DISCUSSION TITLE
하드디스크 접근 시간

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#266] [Computer Architecture]
.

**Comment by 3Juhwan** [comment]
## 하드디스크 접근 시간
하드 디스크에 접근하는 시간은 **탐색 시간**, **회전 지연**, **전송 시간**으로 나뉜다. 

- 탐색 시간: 탐색 시간은 접근하려는 데이터가 저장된 트랙까지 헤드를 이동시키는 시간을 의미한다. 
- 회전 지연: 헤드가 있는 곳으로 플래터를 회전하는 시간을 의미한다. 
- 전송 시간: 하드 디스크와 컴퓨터 간의 데이터를 전송하는 시간을 의미한다. 

*다중 헤드 디스크
- 다중 헤드 디스크는 헤드가 트랙별로 여러 개 달려 있다. 탐색 시간이 0이다. 
- 고정 헤드 디스크라고도 부른다. 

## 프로그래머가 꼭 알아야 할 컴퓨터 시간들

| 항목                                    | 일반적인 시간 범위       |
|-----------------------------------------|-------------------------|
| L1 캐시 참조 시간                        | 0.5 나노초 (ns)        |
| L2 캐시 참조 시간                        | 5 나노초 (ns)       |
| 메모리 참조 시간                         | 7 나노초 (ns)     |
| 메모리에서 1MB를 순차적으로 읽는 시간    | 250 마이크로초 (µs) |
| (하드) 디스크 탐색 시간                  | 10 밀리초 (ms)       |
| (하드) 디스크에서 1MB를 순차적으로 읽는 시간 | 30 밀리초 (ms)     |
| 한 패킷이 캘리포니아에서 네덜란드까지 왕복하는 시간 | 150 밀리초 (ms)    |

**Comment by le2sky** [comment]
**디스크 액세스 시간**

일반적으로 단일-헤드 디스크시스템에서는 특정 데이터 블록(하나 이상의 섹터로 이루어짐)을 읽거나 쓰는 과정은 다음과 같고, 모든 과정이 이루어지는 시간을 **디스크 액세스 시간**이라고한다.

(1) 헤드를 해당 트랙으로 이동시킨다. = 탐색 시간(seek time)
(2) 원하는 데이터가 저장된 섹터가 헤드 아래로 회전되어 올 떄까지 기다린다. = 회전 지연 시간(rotational latency)
(3) 데이터를 전송한다. = 데이터 전송 시간(data transfer time)

```
디스크 액세스 시간 = 탐색 시간 + 회전 지연시간 + 데이터 전송 시간
```

**탐색 시간**
- 기계적인 이동이기 때문에 시동 시간(startup time)이 필요하다.
- 여러 트랙들을 지나가게 되면 적어도 수 ms가 걸리며 거리가 멀면 그 이상도 걸린다.
- 액세스할 데이터들의 지역성(locality)이 높은 경우 헤드의 이동 거리들이 짧아지기 때문에 평균 탐색 시간을 단축 시킬 수 있다.

**회전 지연 시간**
- 회전축의 속도가 7200rpm(분당 회전수)일 경우를 가정하자.
- 위 경우 초당 회전수는 7200 / 60 = 120
- 따라서, 1 회전에 1 /  120 =  0.008초(8ms)가 소요된다.
- 만약 헤드가 해당 트랙에 도착한 순간 원하는 섹터가 헤드 아래에 바로 있다면 회전 지연 시간은 0이다.
- 혹은 원하는 섹터가 직전에 지나갔다면 한 바퀴 더 돌아야 해서 최대 8ms이다.
- 따라서 평균 회전 지연 시간은 4ms가 된다.

**데이터 전송 시간**
- 블록의 크기, 회전 속도, 트랙의 저장 밀도, 버스 전송률 및 제어기 내부 전자회로의 속도에 따라 결정된다.

**예제**
- 디스크 평균 탐색 시간 : 8ms
- 전송률 : 5000 B/s
- 섹터 크기 : 512 Byte
- 회전축 속도 : 5400 rpm
- 제어기 지연시간 :  1ms
